### PR TITLE
Validate `logical` built-ins in JSON schema metaschema

### LINKED ISSUE #348: Add spec test for built-in logical types
https://github.com/recap-build/recap/pull/346/files#r1282108522

Currently the JSON Schema meta schema for recap does not know about [built in logical types](https://recap.build/specs/type/0.2.0/#built-in-logical-types).

A PR to resolve this issue should:

* add a new def for each of the logical types
* add valid test recap specs using the logical types

### LINKED ISSUE #354: Remove `additionalProperties` from 0.2.0.json
I noticed we are prohibiting additional properties from the Recap types in 0.2.0.json metaschema. Additional properties are allowed, though. They're treated as `extra_attrs` in the Python impl. They're essentially ignored (except stuff like default, name)

### PR DESCRIPTION
**Pull request by criccomini**
0.2.0.json now specifies if/then requirements for `logical` types that are built-in (e.g. `build.recap.UUID`, `build.recap.Date`, etc.). If a known logical type is defined, it's validated using the JSON schema metaschema.

If the logical type is unknown, the attributes are ignored and validation passes.

I made a few changes to get this all to work:

1. Using a fixture to open JSON metaschemas and pass them one-by-one into tests. Old way was loading all the schemas into a single list and passing that whole thing into validate and from_dict. This wasn't exactly correct. Error messages are better now, too.
2. Added a `validate()` method to `RecapType`, which is called in `from_dict` for each type. Right now, only `ListType` validation is implemented.
3. Removed a few valid tests that were actaully invalid, and a few invalid tests that were actually valid.

Closes #354
Closes #348

### CODE REVIEW

**Comment by criccomini** [comment]
@mjperrone Could use your eyes on this one.

Notes:

1. I noticed the test_json_schema_spec was loading the entire JSON file and using that as `schema`. This was a list of schemas, actually. It was *kinda* working with Recap's from_dict because from_dict was treating it as a union. But it wasn't really semantically correct. Each `{}` in the list is a separate schema to be validated against the metaschema, and a failure on one shouldn't affect the others. I also feel that the JSON metaschema should define a RecapType at the root, not a list of RecapTypes.

2. I added a `validate` method to make list_field_fixed_no_length fail for Recap`test_recap_rejects_unimplemented_invalid_schemas`. It wasn't failing once I made the changes to (1).

3. I removed `additionalProperties: false`. Recap allows undefined properties. They just get ignored by the spec. They're stored in `extra_attrs` in `RecapType`.

4. I ended up having to do a nested "allOf" for the logical types. The first one is in `RecapType`, here:

https://github.com/recap-build/recap/pull/360/files#diff-67308170bb2e401470c84604fbcf7f2018015232fa32fe798fce77993094c14dR45

The second is in the `RecapLogicalTypes` here:

https://github.com/recap-build/recap/pull/360/files#diff-67308170bb2e401470c84604fbcf7f2018015232fa32fe798fce77993094c14dR392

I'm not sure of a better way to do this; my JSON schema foo is weak. Any ideas?

File: tests/0.2.0.json [hunk 1]
```diff
@@ -155,7 +152,7 @@
         "bytes": {
           "type": "integer",
           "minimum": 1,
-          "maxiumum": 9223372036854775807,
+          "maximum": 9223372036854775807,
```
Comment (mjperrone) [inline tests/0.2.0.json:155]: 😓

Comment (criccomini) [reply tests/0.2.0.json:155]: 😆 ChatGPT actually caught that for me :P

File: tests/0.2.0.json [hunk 2]
```diff
@@ -391,7 +381,246 @@
...
+      "not": {
+        "required": [
+          "alias"
+        ]
       }
+    },
+    "RecapLogicalTypes": {
+      "allOf": [
+        {
+          "if": {
```
Comment (mjperrone) [inline tests/0.2.0.json:394]: TIL about [conditionals in jsonschema](https://json-schema.org/understanding-json-schema/reference/conditionals.html#if-then-else). that's cool! 

I was expecting this to be implemented with `oneOf` and a discriminated union like the other RecapTypes, but I suppose doing it this way allows for non builtin logical types

File: tests/0.2.0.json [hunk 3]
```diff
@@ -391,7 +381,246 @@
...
+              "type": {
+                "const": "int"
+              },
+              "logical": {
+                "const": "build.recap.Timestamp"
+              }
+            },
+            "required": ["type", "logical"]
+          },
+          "$comment": "Not validating timezon field because it's a huge list.",
```
Comment (mjperrone) [inline tests/0.2.0.json:571]:
```suggestion
          "$comment": "Not validating timezone field because it's a huge list.",
```

File: tests/0.2.0.json (hunk 2 above)
Comment (criccomini) [reply tests/0.2.0.json:394]: Haha yea.. that was me a couple days ago :P

Comment (criccomini) [reply tests/0.2.0.json:394]: > I was expecting this to be implemented with oneOf and a discriminated union like the other RecapTypes, but I suppose doing it this way allows for non builtin logical types

Do you think the oneOf way is clearer? What would the "dicriminating union" look like in that case? I'm not totally sold on the if/then.. it feels code-y to me.

File: tests/0.2.0.json (hunk 3 above)
Comment (mjperrone) [reply tests/0.2.0.json]: I spent some time looking for jsonschema definitions of a timestamp field so we could just reference out to that, but I couldn't find one. There is this list... https://github.com/dmfilipenko/timezones.json/blob/master/timezones.json

File: tests/spec/test_json_schema_spec.py [hunk 4]
```diff
@@ -28,73 +32,98 @@ def meta_schema() -> dict:
...
-
-@pytest.mark.parametrize("schema_path", valid_schema_paths)
-def test_valid_schemas(schema_path: str, meta_schema: dict):
-    with open(schema_path, "r") as file:
-        schema = json.load(file)
-        # throws ValidationError if invalid
+
+@pytest.mark.parametrize(
+    "filename,schema",
+    get_schemas(VALID_SCHEMA_DIR, lambda file: file.endswith(".json")),
```
Comment (mjperrone) [inline tests/spec/test_json_schema_spec.py:49]: much nicer

File: tests/spec/test_json_schema_spec.py [hunk 5]
```diff
@@ -28,73 +32,98 @@ def meta_schema() -> dict:
...
-def test_unimplemented_invalid_schemas(schema_path: str, meta_schema: dict):
-    with open(schema_path, "r") as file:
-        schema = json.load(file)
-        with pytest.raises(ValidationError):
-            validate(schema, meta_schema)
+@pytest.mark.parametrize(
+    "filename,schema",
+    get_schemas(
+        INVALID_SCHEMA_DIR,
+        lambda file: file.endswith(".json") and file.endswith(".unimplemented.json"),
```
Comment (mjperrone) [inline tests/spec/test_json_schema_spec.py:97]:
```suggestion
        lambda file: file.endswith(".unimplemented.json"),
```

File: tests/spec/test_json_schema_spec.py [hunk 6]
```diff
@@ -28,73 +32,98 @@ def meta_schema() -> dict:
...
-        schema = json.load(file)
-        with pytest.raises(ValueError):
-            registry = RecapTypeRegistry()
-            recap_type = from_dict(clean_dict(schema), registry)
-            check_aliases(recap_type, registry)
+@pytest.mark.parametrize(
+    "filename,schema",
+    get_schemas(
+        INVALID_SCHEMA_DIR,
+        lambda file: file.endswith(".json") and file.endswith(".unimplemented.json"),
```
Comment (mjperrone) [inline tests/spec/test_json_schema_spec.py:115]:
```suggestion
        lambda file: file.endswith(".unimplemented.json"),
```

File: tests/0.2.0.json (hunk 1 above)
Comment (mjperrone) [reply tests/0.2.0.json:155]: You asked gpt for a review? I should start doing that...

File: tests/0.2.0.json (hunk 2 above)
Comment (mjperrone) [reply tests/0.2.0.json:394]:
```
    "RecapType": {
      "oneOf": [
        {
          "$ref": "#/$defs/RecapNull"
        },
        {
          "$ref": "#/$defs/RecapBool"
        },
```
 `RecapType` is `oneOf` `RecapNull`, `RecapBool`, ... - where `Recap*` looks like
 
```
     "RecapNull": {
      "type": "object",
      "required": ["type"],
      "properties": {
        "type": {
          "type": "string",
          "enum": ["null"]
        },
```

The _union_ is the `oneOf`, and the system _discriminates_ between them based on the `type` field.

Since logical types appear to be subsets of types, you might be able to union the existing `RecapBytes` with the decimal-bytes logical type? I'm not 100% clear on what logical types are and how they function in recap though...



Also, using oneOf vs if-then might yield different error validation outputs that may be more or less user friendly, so that may be a consideration when designing the json schema meta schema.

File: tests/0.2.0.json (hunk 1 above)
Comment (criccomini) [reply tests/0.2.0.json:155]: Haha, nah I asked it to write some of the code :P It's pretty handy for test writing...

File: tests/0.2.0.json (hunk 2 above)
Comment (criccomini) [reply tests/0.2.0.json:394]: I'll poke at it a bit tomorrow, timeboxed to an hour or two. If I can't get something definitively better, I'll move on. Can always change it later.

**Review by mjperrone — APPROVED** [review]

File: tests/0.2.0.json (hunk 2 above)
Comment (mjperrone) [reply tests/0.2.0.json:394]: I'm revisiting this now, as using a pydantic model [generator](https://github.com/koxudaxi/datamodel-code-generator/) from this spec chokes on the `oneOf` `allOf` `if/then`, but it can handle more options on the `oneOf`.

Would you accept a PR changing this format? I'd also need some guidance on how to test it since the tests [hard code](https://github.com/recap-build/recap/blob/main/tests/spec/test_json_schema_spec.py#L23) the published schema.

Comment (criccomini) [reply tests/0.2.0.json:394]: > Would you accept a PR changing this format?

Yes! But I think we need to be careful about making sure that the spec JSON schema continues to express the actual rules. I'm concerned you might find you need to relax the JSON schema in order to make it work with the generator. Doing so would no longer make it a valid spec.

> I'd also need some guidance on how to test it since the tests [hard code](https://github.com/recap-build/recap/blob/main/tests/spec/test_json_schema_spec.py#L23) the published schema.

IIRC, you added [an environment variable](https://github.com/recap-build/recap/blob/main/tests/spec/test_json_schema_spec.py#L28C38-L28C59) for such an occasion... :)

Comment (criccomini) [reply tests/0.2.0.json:394]: WDYT about:

1. Adding the `generator` task to the CI (it should fail at first)
2. Updating the JSON schema (without updating any tests)
3. Validating that the generator now works

I think if  you can do this without having to modify the tests, it *should* be a safe spec change... We'll still have to sniff around since the tests don't cover every single edge case. But it should be pretty close...

Comment (criccomini) [reply tests/0.2.0.json:394]: @mjperrone checking in here

Comment (mjperrone) [reply tests/0.2.0.json:394]: sorry, other priorities came up. I think I have time to at least make the initial PR changing the definition method, not sure I'll have the time to validate it

Comment (mjperrone) [reply tests/0.2.0.json:394]: [github.com/recap-build/recap-website/pull/17](https://github.com/recap-build/recap-website/pull/17): `use discriminated union (oneOf) instead of if/then to define logical types` (+156, −209)
